TestGetLines test fails in i686

```
=== RUN   TestGetLines
    reader_test.go:59: Testing file: ../sample-files/8-bit-color.txt...
    reader_test.go:59: Testing file: ../sample-files/compressed.txt.bz2...
    reader_test.go:59: Testing file: ../sample-files/compressed.txt.gz...
    reader_test.go:159: Not testing xz compressed file, xz not found in $PATH:  ../sample-files/compressed.txt.xz
    reader_test.go:59: Testing file: ../sample-files/dos.txt...
--- FAIL: TestGetLines (0.02s)
panic: runtime error: slice bounds out of range [:-2147483640] [recovered]
	panic: runtime error: slice bounds out of range [:-2147483640]
```

Full log of test run at https://github.com/void-linux/void-packages/actions/runs/5371471390/jobs/9744251742#step:7:1805

--- a/.dockerignore
+++ b/.dockerignore
@@ -1,0 +1,1 @@
+/releases

--- a/.github/workflows/linux-ci.yml
+++ b/.github/workflows/linux-ci.yml
@@ -15,7 +15,7 @@ jobs:
       - name: Set up Go
         uses: actions/setup-go@v3
         with:
-          go-version-file: 'go.mod'
+          go-version-file: "go.mod"
 
       # golangci-lint is required by test.sh. Latest version here if you want
       # to bump it, version number is at the end of the "curl | sh"
@@ -25,3 +25,4 @@ jobs:
         run: curl -sSfL https://raw.githubusercontent.com/golangci/golangci-lint/master/install.sh | sh -s -- -b "$(go env GOPATH)"/bin v1.51.2
 
       - run: ./test.sh
+      - run: GOARCH=386 ./test.sh

--- a/Dockerfile-test-386
+++ b/Dockerfile-test-386
@@ -1,0 +1,13 @@
+# Run the tests in 32 bit mode:
+# docker build . -f Dockerfile-test-386
+
+FROM golang:1.20
+
+WORKDIR /moar
+
+COPY go.mod go.sum ./
+RUN go mod download
+
+COPY . .
+
+RUN GOARCH=386 go test -v ./...

--- a/README.md
+++ b/README.md
@@ -184,6 +184,9 @@ Run tests:
 ./test.sh
 ```
 
+To run tests in 32 bit mode, either do `GOARCH=386 ./test.sh` if you're on
+Linux, or `docker build . -f Dockerfile-test-386` (tested on macOS).
+
 Run microbenchmarks:
 
 ```bash

--- a/m/reader.go
+++ b/m/reader.go
@@ -5,6 +5,7 @@ import (
 	"bytes"
 	"fmt"
 	"io"
+	"math"
 	"os"
 	"os/exec"
 	"path"
@@ -517,6 +518,18 @@ func (r *Reader) GetLines(firstLineOneBased int, wantedLineCount int) (*InputLin
 	return r.getLinesUnlocked(firstLineOneBased, wantedLineCount)
 }
 
+func nonWrappingAdd(a int, b int) int {
+	if b < 0 {
+		panic(fmt.Sprintf("b must be >= 0, was: %d", b))
+	}
+
+	if a > math.MaxInt-b {
+		return math.MaxInt
+	}
+
+	return a + b
+}
+
 func (r *Reader) getLinesUnlocked(firstLineOneBased int, wantedLineCount int) (*InputLines, overflowState) {
 	if firstLineOneBased < 1 {
 		firstLineOneBased = 1
@@ -535,7 +548,7 @@ func (r *Reader) getLinesUnlocked(firstLineOneBased int, wantedLineCount int) (*
 	}
 
 	firstLineZeroBased := firstLineOneBased - 1
-	lastLineZeroBased := firstLineZeroBased + wantedLineCount - 1
+	lastLineZeroBased := nonWrappingAdd(firstLineZeroBased, wantedLineCount-1)
 
 	if lastLineZeroBased >= len(r.lines) {
 		lastLineZeroBased = len(r.lines) - 1

--- a/m/reader_test.go
+++ b/m/reader_test.go
@@ -171,9 +171,11 @@ func TestGetLines(t *testing.T) {
 			continue
 		}
 
-		testGetLines(t, reader)
-		testGetLineCount(t, reader)
-		testHighlightingLineCount(t, file)
+		t.Run(file, func(t *testing.T) {
+			testGetLines(t, reader)
+			testGetLineCount(t, reader)
+			testHighlightingLineCount(t, file)
+		})
 	}
 }
 

--- a/m/reader_test.go
+++ b/m/reader_test.go
@@ -56,8 +56,6 @@ func testGetLineCount(t *testing.T, reader *Reader) {
 }
 
 func testGetLines(t *testing.T, reader *Reader) {
-	t.Logf("Testing file: %s...", *reader.name)
-
 	lines, _ := reader.GetLines(1, 10)
 	if len(lines.lines) > 10 {
 		t.Errorf("Asked for 10 lines, got too many: %d", len(lines.lines))
@@ -157,6 +155,14 @@ func TestGetLines(t *testing.T) {
 			_, err := exec.LookPath("xz")
 			if err != nil {
 				t.Log("Not testing xz compressed file, xz not found in $PATH: ", file)
+				continue
+			}
+		}
+
+		if strings.HasSuffix(file, ".bz2") {
+			_, err := exec.LookPath("bzip2")
+			if err != nil {
+				t.Log("Not testing bzip2 compressed file, bzip2 not found in $PATH: ", file)
 				continue
 			}
 		}
@@ -306,9 +312,15 @@ func testCompressedFile(t *testing.T, filename string) {
 
 func TestCompressedFiles(t *testing.T) {
 	testCompressedFile(t, "compressed.txt.gz")
-	testCompressedFile(t, "compressed.txt.bz2")
 
-	_, err := exec.LookPath("xz")
+	_, err := exec.LookPath("bzip2")
+	if err == nil {
+		testCompressedFile(t, "compressed.txt.bz2")
+	} else {
+		t.Log("WARNING: bzip2 not found in path, not testing automatic bzip2 decompression")
+	}
+
+	_, err = exec.LookPath("xz")
 	if err == nil {
 		testCompressedFile(t, "compressed.txt.xz")
 	} else {

